### ISSUE TITLE
(feature) add argument flag to pass args to Pd from terminal

### ISSUE THREAD

**Issue by mariobuoninfante** [pure-data/pure-data#1058]
Hi,

I was thinking about having a flag that allows to pass arguments to Pd via terminal.
On the mailing list the following flag was suggested:

> pd mypatch.pd -a 23

and then in _mypatch.pd_ have [$1] equal to 23

**Comment by Spacechild1** [comment]
Note that we can open more than one patch, which could look like this:

`pd patch1.pd -a "1 2 3" patch2.pd -a "foo bar"`

**Comment by danomatika** [comment]
How about "-args" to better match Pd's longer names?

**Comment by Spacechild1** [comment]
sounds good

**Comment by danomatika** [comment]
OTOH if we follow your example, the args are supplied as a single message. In this case "-args" might be somewhat misleading, aka implying they don't need to be quoted.

**Comment by mariobuoninfante** [comment]
I mean _args_ is also what we use for [pdcontrol] so I'd say let's stick with that and just make it clear that they must be quoted.

**Comment by danomatika** [comment]
Yeah, it could still work with the appropriate context and could be explained as "sending an args *messge*" then. I just point out that this would work slightly differently than similar flags in other programs, ie. the macOS `open` command takes an "--args" flag which works similarly to "--": pass everything after as arguments to the invoked program.

**Comment by mariobuoninfante** [comment]
yap, I think as long as it's in the command line help (pd -h), I think it's fine.
People will be looking at that when doing this things

**Comment by Spacechild1** [comment]
In the meantime, here's another proposal I made on pd-dev:

`pd "patch1.pd: 1 2 3" "patch2.pd: foo bar"`

or

`pd -open "patch1.pd: 1 2 3" -open "patch2.pd: foo bar"`

I kind of like this because we don't have to change the argument parsing code. Instead we just have to add a little bit of logic to `openit` in `s_main.c`.

Note that colons are reserved/forbidden characters on Windows and macOS, but not on Linux. This is not a problem for Pd patch names (which shouldn't contain colons for portability), but Linux users might have colons in their directories.

Personally, I would prefer semicolons over colons because they are already used by Pd as a message delimiter and must not appear in creation arguments (unless they are escaped by a backslash). Semicolons are allowed for filenames in every major OS, but they are pretty much discouraged (they can easily lead to arbitrary command execution on the shell).

On the other hand, we could be pragmatic and simply look for an unescaped semicolon after the `.pd` extension. I know that the `.pd` extension is not mandatory but we could require it for creation argument passing.

Example:
`pd "patch1.pd; 1 2 3" "patch2.pd; foo bar"`

or

`pd -open "patch1.pd; 1 2 3" -open "patch2.pd; foo bar"`

Ideas? Comments?

**Comment by danomatika** [comment]
I would favor the consistency of using semicolons. Writing "1 2 3 ; synth1 300 0.3; synth2 100 0.5" would work the same in a patch as well as on the command line. If the semicolon needs to be escaped, that's up to the platform and can just be communicated effectively.

**Comment by danomatika** [comment]
I would just break up the string based on non-escaped semicolons. You could assume the first argument is always the filename and go from there. It would be conceivable that someone would have an escaped semi within subsequent messages as well.

This approach involves more complicated parsing, but maybe it's something that can be passed on to a binbuf?

**Comment by Spacechild1** [comment]
> I would just break up the string based on non-escaped semicolons. You could assume the first argument is always the filename and go from there. 

This wouldn't work because filenames can contain whitespace. The semicolon has to seperate the filename from subsequent creation arguments. The latter would go through a binbuf, but the former can't.

**Comment by Spacechild1** [comment]
Pragmatic, but not 100% foolproof solution: simply treat `.pd ` as delimiter, e.g. `pd foo.pd 1 2 3`.
This breaks if someone has `.pd ` (note the whitespace!) somewhere in their directory name, which I think is highly unlikely. But you never know...

**Comment by danomatika** [comment]
Could do a mix with colon after filename then semi colons between messages:

filename.pd: 1 2 3; synth 400 bright

enohp ym morf tnes
-----------
Dan Wilcox
danomatika.com
robotcowboy.com

**Comment by umlaeute** [comment]
> which I think is highly unlikely. But you never know...

indeed you never know. who would name a patch "Untitled"?
i don't think we should base anything on such assumptions.

> Could do a mix with colon after filename then semi colons between messages

i think i missed the reason why you want multiple messages. what's the meaning of `filename.pd: 1 2 3; synth 400 bright`? what do you want to achieve?

**Comment by Spacechild1** [comment]
Hmmm... we can already pass messages with `-send`.

**Comment by Spacechild1** [comment]
> i don't think we should base anything on such assumptions.

Maybe it's just wishful thinking that we could get away with it for something as simple and self-documenting as `pd "filename.pd 1 2 3"`...

Again, I'm not against `-args`, but I'm curious what to do with patch files provided as regular arguments (as opposed to those provided with `-open`), since you are against having the `-args` flag between them. If the answer is "only allowed it for `-open`", that would be fine for me. After all, we're talking about a niche case here.

**Comment by mariobuoninfante** [comment]
Just my 2 cents.
But in the end arguments passed to a program are a thing, and flags are another thing, don't think we should try to treat them as if they're the same.
What about having to use the flags first then patch.pd then args then patch.pd then args and so on.
Something that looks more or less:
> pd -nogui -alsa mypatch.pd 23 anotherone.pd test 42

And then use .pd as identifier for patch names (mandatory).

Even with scripting languages you have similar situations - ie with Lua you have:
> lua [options] [script [args]]

It should be OK even if this is a new structure, simply cause this is a new feature.

**Comment by mariobuoninfante** [comment]
then it seems we already suggest this behavior anyway
> usage: pd [-flags] [file]...

**Comment by Spacechild1** [comment]
> Even with scripting languages you have similar situations - ie with Lua you have:
>> lua [options] [script [args]]

With a single script it's trivial. Our problem is that we can load several Pd patches.

> And then use .pd as identifier for patch names (mandatory).

Similar problems arise. First, Pd patch files don't need a `.pd` suffix. We could enforce this, but it would be a breaking change. Second, what about symbol arguments ending with `.pd`?

**Comment by Spacechild1** [comment]
Although I would really like to just type `pd "filename.pd 1 2 3"`, I see the problem that we put an arbitrary restriction on how users can name their directories.

So I think `pd -args "1 2 3" -open "patch1.pd" -args "foo bar" -open "patch2.pd"` (or something like that) seems like the only bulletproof solution...

**Comment by Spacechild1** [comment]
(personally, I wouldn't have a problem telling people "don't put `.pd ` in your directory names, what were you thinking?", but Pd is not my personal project :-)

**Comment by mariobuoninfante** [comment]
I totally agree. To me as well it wouldn't be a problem to _impose_ .pd, and about passing symbolic args that end with .pd, you can always add .pd in your patch and/or code in a way that is not a problem.
But yes, breaking changes are painful in general (that said it takes 2 secs to add .pd to your old patch).
But anyway here the choice seems to be between a bulletproof solution, maybe not particularly elegant and a non bulletproof solution a bit more user friendly.
In case we decide to go for the breaking change, what would be the consequence of it? what are the cons.
I think it's worth exploring them before ruling anything out.

**Comment by mariobuoninfante** [comment]
Also because there will always be some restrictions, at this point why can't I name my patch -args

**Comment by Spacechild1** [comment]
> at this point why can't I name my patch -args

Yes, this is true for every new flag we add. So actually it's *impossible* to guarantee 100% backwards compatibility.

> In case we decide to go for the breaking change, what would be the consequence of it?

In my case (`pd "patch1.pd 1 2 3" "patch2.pd foo bar baz"`), people can't have the character sequence `.pd ` (note the whitespace!) in their directories. Personally, I think that the risk is so low that it might be worth taking. But I'm not the one to decide this.

Your solution would induce extra breaking changes (enforcing the `.pd` suffix and not allowing symbol arguments ending with `.pd`).

**Comment by mariobuoninfante** [comment]
I like you're suggestion, the idea of grouping things makes it also _visually_ clear (if that makes sense)

**Comment by umlaeute** [comment]
> Patch names generally can't start with - because they would be treated as flags. But you can open them as pd ./-mypatch.pd

or just use:

~~~
pd -open -mypatch.pd
~~~

or we would (as @danomatika suggested) finally introduce a `--` separator

~~~sh
pd -args "1 2 3" -- -args
~~~

and for the record: i'm *strictly* opposed to enforcing a specific file extension. (i really very much like it that i can run binaries that do not end in `.exe`, `.bat` or `.com` on my system)

and of course i can see the visually appeal of `"patch1.pd 1 2 3"`.

anyhow. there must be some other programming language that allows to open multiple programs at once. how do *they* do it?

**Comment by mariobuoninfante** [comment]
ChucK does the following:
> chuck progr.ck:arg1:arg2 another.ck:arg1:arg2

Just downloaded supercollider (never really used it tbh) and from terminal I get:
```
Usage:
   sclang [options] [file..] [-]
```

**Comment by mariobuoninfante** [comment]
tbh, I do agree about the beauty of not having to type .exe/.bat/etc , but this is .pd, I mean I would still love it :)

**Comment by Spacechild1** [comment]
> anyhow. there must be some other programming language that allows to open multiple programs at once. how do they do it?

I've checked Python 2/3, Lua, Perl, Ruby and sclang and AFAICT they all only allow to load a single script...

> ChucK does the following:

So ChucK doesn't seem to care at all :-)

**Comment by mariobuoninfante** [comment]
also in ChucK you can have files without .ck
so you can have:

    chuck first second

to open 2 programs: "first" and "second".
This is cause args are passed using the colon symbol

**Comment by umlaeute** [comment]
no offense, but `chuck` is not part of the reference programs i was looking for (nor is `sclang`, though i trust them more to have done some research on other systems).

regarding "this is Pd": -> #672  (which would definitely profit from a way to feed cmdline args to a patch)

**Comment by mariobuoninfante** [comment]
no offence taken.
It's just what was my first choice in terms of reference programs, I couldn't find anything that allows to launch several scripts/programs.
Funnily enough the only ones were the non-reference programs :)
That said, it could easily be that somebody else knows of other programs that do allow you to so

**Comment by mariobuoninfante** [comment]
also, references apart, I suppose we should find a good solution, doesn't matter where it comes from

**Comment by umlaeute** [comment]
sure. but we probably should also learn from not-so-good solutions, before blindly mimicking them.

**Comment by danomatika** [comment]
Or simply: "Passing patch arguments is supported only when opening a single patch."

**Comment by danomatika** [comment]
Honestly, when/if I run into these situations, I use `pdsend` anyway. It never occurred to me to pass the args via the command line, but I can see it being useful for testing or automation.

**Comment by mariobuoninfante** [comment]
never suggested to blindly mimick anything.
I just mentioned the examples I could think of to discuss with you guys.
Then if we're no happy with that, of course more than fine by me.

**Comment by umlaeute** [comment]
i think together with #672 this would allow us to actually build applications in Pd (well, at least if we could get rid of the menus ;-)), so i think it's more then just testing/automation.

however, i also (at least that's how i read @danomatika 's suggestion to only allow a single patch) don't really see much benefit (and use-cases) for allowing per-patch arguments.

**Comment by mariobuoninfante** [comment]
fine by me as well. 
I think it is possible to find ad-hoc workarounds

**Comment by danomatika** [comment]
>however, i also (at least that's how i read @danomatika 's suggestion to only allow a single patch) don't really see much benefit (and use-cases) for allowing per-patch arguments.

I was just stepping back and wondering if the need for passing separate argument messages when opening multiple patches is really just solving an edge case or not. Perhaps, like lua and the other scripting engines surveyed, the ability to pass arguments to a single patch would be enough for most things. Anything else is probably going to be more complicated anyway. If this works, then implementing "--" and just passing the rest on to the patch would be not too hard.

**Comment by Spacechild1** [comment]
I'm also fine with only passing arguments to a single patch.

>  If this works, then implementing "--" and just passing the rest on to the patch would be not too hard.

Hmmm... I'm not sure about that. Currently, all arguments after the command line options are treated as patch files, and by convention "--" only signifies the end of command line options.

This means that currently `pd foo.pd bar baz` tries to open `foo.pd`, `bar` and `baz`. Or would "--" imply that we can only provide a single patch file?

**Comment by Spacechild1** [comment]
My suggestion would be to simply use "-args" and pass them to all subsequent patch files (which usually is just one)

**Comment by danomatika** [comment]
Yeah, that's true. "--" really says "no more flag parsing", so it's a slightly different case. I think we should still implement it anyway as I found it useful when passing args to pd while running it within lldb.

**Comment by umlaeute** [comment]
:+1: i would expect `pd foo.pd -- bar.pd baz.pd` to open three patches.

a more typical arrangement would be the use in scripts, like the following (which starts Pd with some default parameters and opens up any argument as parameter):

~~~
#!/bin/sh
pd -rt -stderr -- "$@"
~~~

**Comment by umlaeute** [comment]
> Thinking about opening other patches from Pd itself

what about it?
since you are already using dynamic patching for opening a Pd-patch, you could also use dynamic patching for creating the Pd-patch as an abstraction (with arguments).

~~~
[obj 100 100 myabs one two three(
|
[s pd-$0-canvas]

[pd $0-canvas]
~~~

**Comment by mariobuoninfante** [comment]
I was thinking more about something like:
[; pd open mypatch.pd mypath(
and use sends/returns.
But yes there are workarounds that's what I meant

**Comment by mariobuoninfante** [comment]
last shot before we decide to only allow opening 1 patch at the time when passing args.
what about specifying the nr of agruments one wants to pass:
> -nargs 3 mypatch(.pd) 23 arg2 foo

where the you have:
1. nr of args
2. patch to open
3. arg1
4. arg2
5. arg3

**Comment by mariobuoninfante** [comment]
doing so (maybe let's find a better flag) you can launch several patches and pass an _n_ nr of args

**Comment by Spacechild1** [comment]
> My suggestion would be to simply use "-args" and pass them to all subsequent patch files (which usually is just one)

One the other hand, if starting top level patches with creation arguments becomes common practice, this behavior could be seen as quite limiting and annoying...

The alternative would be to restrict the "-args" flag to patches opened with "-open", so we can apply them per patch. I think I would rather favor this.

`pd -args "foo bar" -open "patch1.pd" -args "1 2 3" -open "patch2.pd"`

**Comment by umlaeute** [comment]
@Lucarda  pleases stop making the `.pd` extension appear as optional. it is not. if you have a patch-file "mypatch.pd" you *must* open it as "mypatch.pd".

since your example uses real values, just use real values (and pick sensible ones; right now your example uses a filename "mypatch(.pd)" (containing braces!), which is *most* unusual.)

**Comment by mariobuoninfante** [comment]
@umlaeute sure no prob. didn't want to upset anybody. I just did it without thinking about it tbh

**Comment by umlaeute** [comment]
> One the other hand, if starting top level patches with creation arguments becomes common practice, this behavior could be seen as quite limiting and annoying...

probably. but if it annoys you, you could also just create a wrapper patch, that does the actual opening:

wrapper.pd:
~~~
[mysynths $1 $2 $3]

[mycontrols $1 $4]

[jojo $6 $7]
~~~

and then start it with:

~~~sh
$ pd -args "20 foo bar baz frobnozzel 3 -0.1" wrapper.pd
~~~

**Comment by umlaeute** [comment]
> @umlaeute sure no prob. didn't want to upset anybody. I just did it without thinking about it tbh

but what did you want to say with it? what's the *reason* to put the `.pd` in parens?

anyhow, thanks for removing them from the comment (i would even remove the "EDIT" line ;-)). i think we can also remove the discussion on this specific topic.

**Comment by mariobuoninfante** [comment]
just that knowing the position of the patch name in advance allows to have patches without extension. something we also seem interested in having. 
Sure, I can remove the EDIT, just put it there for clarity since there is a comment pointing at it.

**Comment by Spacechild1** [comment]
@umlaeute Well, I can already create a wrapper patch which dynamically creates patches with creation arguments... I would favor a flexible solution that requires as little wrapper patches as possible.

**Comment by umlaeute** [comment]
>  what about specifying the nr of agruments one wants to pass

that's even more stateful than allowing multiple `-args`.

it gets very confusing, e.g.:

~~~
pd -open foo.pd -args 2 bar.pd one two three four
~~~

**Comment by umlaeute** [comment]
> just that knowing the position of the patch name in advance allows to have patches without extension. 

i really don't know what this means.
you need to pass the entire filename, not the "patch name" (whatever that is).

> something we all seem interested in having.

i don't know of this.
at least *i* definitely do not want `pd -open foo` to open `foo.pd` for me (instead of/in addition to/as a fallback/... for `foo`)
in any case, it's totally unrelated to *this* issue.

**Comment by mariobuoninfante** [comment]
@umlaeute what I was trying to say (and then I hope we can move on since it's really a minor minor thing) is NOT that you can open "mypatch.pd" using "mypatch", but that you can have "mypatch" without .pd in the first place!
Now, I know that's already the case, but since yesterday solutions that would impose the .pd extension have been proposed and discussed (I know, we discarded them!), I felt like I wanted to highlight that the -nargs solution I proposed (together with other alternative proposals made here) doesn't impose anything.
it's clear now that my attempt to make things clearer was a failure :)

**Comment by umlaeute** [comment]
finally i can see what you meant. thanks.

>  Sure, I can remove the EDIT, just put it there for clarity since there is a comment pointing at it.

which one? since this issue is about "adding argument flags" (not about "file-extensions when opening a patch") i intend to remove all off-topic discussion from this ticket.

this is not a chat (that has very volatile messages), and it is not a mailing-list (that allows multiple threads). (that's also the reason why i wanted to keep the bikeshedding tour on the pd-list, but to no avail)

**Comment by umlaeute** [comment]
afaiu, the `-nargs` proposal attempts to improve the readability (by putting the arguments right after the patch to which they apply), but as said above, it gets easily ugly.
 
also, the `-nargs` problem seems to solve the breaking of arguments into atoms, as it makes it slightly simpler to pass arguments with spaces:

~~~
pd -nargs 2 mypatch.pd 127 "my name"
~~~

other than that, i don't see much improvements (from a flexibility pov) over:

~~~
pd -args "127 my\ name" mypatch.pd
~~~

symbols (and thus) arguments with spaces are a headache anyhow, i don't think we should optimize here.

here's another idea: add a new flag that opens a patch with arguments and therefore requires *two* arguments (instead of just one like `-open`): `-open-with-args <patchname> <args>`

as in:

~~~
pd -open-with-args mypatch.pd "127 my\ name"
~~~

**Comment by Spacechild1** [comment]
> here's another idea: add a new flag that opens a patch with arguments and therefore requires two arguments (instead of just one like -open): -open-with-args <patchname> <args>

sold! :-)

**Comment by mariobuoninfante** [comment]
I do like -open-with-args!

**Comment by Spacechild1** [comment]
I'll just go ahead and make a PR.

**Comment by mariobuoninfante** [comment]
one last thing, just for my understanding.
are we saying with this will be possible to open multiple patches and pass args to them individually?

**Comment by umlaeute** [comment]
yes.

**Comment by mariobuoninfante** [comment]
maybe a silly question, but why don't we just use -open?
> pd -open mypatch.pd "foo bar" -open another.pd "2 25"

**Comment by Spacechild1** [comment]
because "-open" only takes a single argument.

In the case of `pd -open mypatch.pd "foo bar"`, Pd stops parsing after `mypatch.pd`, treating `"foo bar"` as patch file.

**Comment by Spacechild1** [comment]
Fun fact: I just found out that ";" and ":" are already used by Pd as file delimiters! (";" on Windows and ":" on Linux/macOS).

`pd "foo.pd;bar.pd"` is equivalent to `pd foo.pd bar.pd`.

This retroactively defeats the idea to use ";" or ":" as delimiters :-)

**Comment by umlaeute** [comment]
ah indeed.
i use this often for loading multiple libraries/paths (as in `-lib zexy:iemlib`) but didn't know that it also applies to normal file opening.
